Sources/SwiftSyntax/BumpPtrAllocator.swift

```diff
@@ -0,0 +1,113 @@
+//===---------- BumpPtrAllocator.swift - Bump-pointer Allocaiton ----------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// Bump-pointer allocation.
+class BumpPtrAllocator {
+  typealias Slab = UnsafeMutableRawBufferPointer
+
+  static private var SLAB_SIZE: Int = 4096
+  static private var GLOWTH_DELAY: Int = 128
+  static private var SLAB_ALIGNMENT: Int = 8
+
+  private var slabs: [Slab]
+  private var currentPtr: UnsafeMutableRawPointer?
+  private var endPtr: UnsafeMutableRawPointer?
+  private var customSizeSlabs: [Slab]
+  private(set) var totalSizeAllocated: Int
+
+  init() {
+    slabs = []
+    currentPtr = nil
+    endPtr = nil
+    customSizeSlabs = []
+    totalSizeAllocated = 0
+  }
+
+  deinit {
+    /// Deallocate all memory.
+    totalSizeAllocated = 0
+    currentPtr = nil
+    endPtr = nil
+    while !slabs.isEmpty {
+      slabs.removeLast().deallocate()
+    }
```

Comment:
What’s the reason that you need to remove the slabs from the array yourself and can’t just do the following?
```swift
for slab in slabs {
  slab.deallocate()
}
```

Reply:
I just don't want to keep the deallocated slabs in the array even in `deinit`. Retting `startPtr` et al, is for the same reason. I don't want to keep any dangling pointers.

Reply:
OK, fair enough. It seem a little too cautious to me but I’m fine with it.

Reply:
I mean, another way to write this and maintain the style is 

```
while let slab = slabs.popLast() {
  slab.deallocate()
}
```

Reply:
Updated to `popLast()`